File: core/canvas/text.go
package canvas

import (
	"fmt"
	"io/ioutil"
	"math"
	"strings"

	"github.com/fogleman/gg"
	"github.com/golang/freetype/truetype"
	"golang.org/x/image/font"
	"golang.org/x/mobile/asset"
)

const LINE_SPACING = 1

func FitText(text string, fontFamily string, width float64, height float64) float64 {
	var low, high float64 = 1, height
	dc := gg.NewContext(int(width), int(height))

	fontSize := low
	for low <= high {
		mid := math.Floor((high + low) / 2)
		font := loadFont(dc, fontFamily, mid)
		multilineText := strings.Join(dc.WordWrap(text, width), "\n")
		mWidth, mHeight := dc.MeasureMultilineString(multilineText, LINE_SPACING)
		font.Close()
		if mWidth <= width && mHeight <= height {
			fontSize = mid
			low = mid + 1
		} else {
			high = mid - 1
		}
	}

	return fontSize
}

var fonts = make(map[string]*truetype.Font)

func loadFont(dc *gg.Context, name string, fontSize float64) font.Face {
	if _, ok := fonts[name]; !ok {
		file, err := asset.Open(fmt.Sprintf("fonts/%s.ttf", name))
		if err != nil {
			panic(err)
		}
		raw, err := ioutil.ReadAll(file)
		if err != nil {
			panic(err)
		}
		font, err := truetype.Parse(raw)
		if err != nil {
			panic(err)
		}
		fonts[name] = font
	}
	face := truetype.NewFace(fonts[name], &truetype.Options{
		Size: float64(fontSize),
	})
	dc.SetFontFace(face)
	return face
}

File: core/src/lib/text.ts
import { CanvasRenderingContext2D, createCanvas } from "canvas"

// https://github.com/fogleman/gg/blob/0403632d5b905943a1c2a5b2763aaecd568467ec/wrap.go#L28
export function wrap(
    context: CanvasRenderingContext2D,
    text: string,
    width: number
) {
    const result: string[] = []
    text.split("\n").forEach((line) => {
        const fields = line.split(" ").map((w) => w + " ")
        if (fields.length % 2 === 1) {
            fields.push("")
        }
        let x = ""
        for (let i = 0; i < fields.length; i += 2) {
            const w = context.measureText(x + fields[i]).width
            if (w > width) {
                if (x === "") {
                    result.push(fields[i])
                    continue
                } else {
                    result.push(x)
                    x = ""
                }
            }
            x += fields[i] + fields[i + 1]
        }
        if (x !== "") {
            result.push(x)
        }
    })
    return result.map((s) => s.trim()).join("\n")
}

export function measureText(
    context: CanvasRenderingContext2D,
    text: string
) {
    const metrics = context.measureText(text)
    return {
        width: metrics.width,
        // @ts-ignore
        height: metrics.emHeightAscent + metrics.emHeightDescent
    }
}

export function fittext({ text, font, width, height }: {
    text: string,
    font: string,
    width: number,
    height: number
}) {
    let low = 1, high = height
    const dc = createCanvas(width, height).getContext("2d")
    let fontSize = low
    while (low <= high) {
        const mid = Math.floor((high + low) / 2)
        dc.font = `${mid}px ${font}`
        const wrapped = wrap(dc, text, width)
        const m = measureText(dc, wrapped)
        if (m.width <= width && m.height <= height) {
            fontSize = mid
            low = mid + 1
        } else {
            high = mid - 1
        }
    }
    return {
        fontSize,
        text: wrap(dc, text, width)
    }
}
